File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/StatisticsViewModel.java
package com.platdmit.simplecloudmanager.app.vm;

import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.platdmit.simplecloudmanager.app.helpers.charts.serverValueFormatter;
import com.platdmit.simplecloudmanager.domain.models.ComplexChartsData;
import com.platdmit.simplecloudmanager.domain.models.Statistic;
import com.platdmit.simplecloudmanager.domain.repo.ServerStatisticsRepo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class StatisticsViewModel extends BaseViewModel {
    private static final String TAG = StatisticsViewModel.class.getSimpleName();
    private ServerStatisticsRepo mServerStatisticsRepo;

    private MutableLiveData<ComplexChartsData> mCpuDataCharts = new MutableLiveData<>();
    private MutableLiveData<ComplexChartsData> mRamDataCharts = new MutableLiveData<>();

    public StatisticsViewModel(ServerStatisticsRepo serverStatisticsRepo, long serverId) {
        mServerStatisticsRepo = serverStatisticsRepo;
        mCompositeDisposable.add(mServerStatisticsRepo.getServerStatistics(serverId).subscribe(this::generateChart));
    }

    private void generateChart(List<Statistic> statistics) {
        Completable.fromAction(() -> {
            List<Entry> CpuEntries = new ArrayList<Entry>();
            List<Entry> RamEntries = new ArrayList<Entry>();
            Map<Float, String> titles = new HashMap<>();
            for (int i = 0; i < statistics.size(); i++) {
                CpuEntries.add(new Entry(i+1, statistics.get(i).getCpuVal()));
                RamEntries.add(new Entry(i+1, statistics.get(i).getRamVal()));
                titles.put((float)i+1, statistics.get(i).getTime());
            }
            LineDataSet dataSetRam = new LineDataSet(RamEntries, "");
            mRamDataCharts.postValue(new ComplexChartsData(new LineData(dataSetRam), new serverValueFormatter(titles)));

            LineDataSet dataSetCpu = new LineDataSet(CpuEntries, "");
            mCpuDataCharts.postValue(new ComplexChartsData(new LineData(dataSetCpu), new serverValueFormatter(titles)));
        }).observeOn(Schedulers.newThread())
                .doOnError(Throwable::printStackTrace)
                .subscribe();
    }

    public LiveData<ComplexChartsData> getCpuDataLiveData() {
        return mCpuDataCharts;
    }
    public LiveData<ComplexChartsData> getRamDataLiveData() {
        return mRamDataCharts;
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/app/vm/StatisticsViewModel.kt
package com.platdmit.simplecloudmanager.app.vm

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.platdmit.simplecloudmanager.app.helpers.charts.ServerValueFormatter
import com.platdmit.simplecloudmanager.domain.models.ComplexChartsData
import com.platdmit.simplecloudmanager.domain.models.Statistic
import com.platdmit.simplecloudmanager.domain.repo.ServerStatisticsRepo
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.schedulers.Schedulers
import java.util.*
import kotlin.collections.ArrayList

class StatisticsViewModel(
        private val mServerStatisticsRepo: ServerStatisticsRepo,
        serverId: Long
) : BaseViewModel() {
    private val mCpuDataCharts = MutableLiveData<ComplexChartsData>()
    private val mRamDataCharts = MutableLiveData<ComplexChartsData>()

    init {
        mCompositeDisposable.add(mServerStatisticsRepo.getServerStatistics(serverId).subscribe { generateChart(it) })
    }

    private fun generateChart(statistics: List<Statistic>) {

        Completable.fromAction {
            val cpuEntries: MutableList<Entry> = ArrayList()
            val ramEntries: MutableList<Entry> = ArrayList()
            val titles: MutableMap<Float, String> = HashMap()

            for ((i, v) in statistics.withIndex()) {
                cpuEntries.add(Entry((i + 1).toFloat(), v.cpuVal))
                ramEntries.add(Entry((i + 1).toFloat(), v.ramVal))
                titles[i.toFloat() + 1] = v.time
            }

            mRamDataCharts.postValue(
                    ComplexChartsData(
                            LineData(LineDataSet(ramEntries, "")),
                            ServerValueFormatter(titles)
                    )
            )
            mCpuDataCharts.postValue(
                    ComplexChartsData(
                            LineData(LineDataSet(cpuEntries, "")),
                            ServerValueFormatter(titles)
                    )
            )

        }.observeOn(Schedulers.newThread())
                .doOnError { it.printStackTrace() }
                .subscribe()
    }

    val cpuDataLiveData: LiveData<ComplexChartsData>
        get() = mCpuDataCharts

    val ramDataLiveData: LiveData<ComplexChartsData>
        get() = mRamDataCharts

    companion object {
        private val TAG = StatisticsViewModel::class.java.simpleName
    }
}